extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -584,4 +858,95 @@ public TopicPartition apply(Integer n)
         )
     );
   }
+
+  private Set<Integer> assignPartitionsAndSeekToNext(KafkaConsumer consumer, String topic)
+  {
+    // Initialize consumer assignment.
+    final Set<Integer> assignment = Sets.newHashSet();
+    for (Map.Entry<Integer, Long> entry : nextOffsets.entrySet()) {
+      final long endOffset = endOffsets.get(entry.getKey());
+      if (entry.getValue() < endOffset) {
+        assignment.add(entry.getKey());
+      } else if (entry.getValue() == endOffset) {
+        log.info("Finished reading partition[%d].", entry.getKey());
+      } else {
+        throw new ISE(
+            "WTF?! Cannot start from offset[%,d] > endOffset[%,d]",
+            entry.getValue(),
+            endOffset
+        );
+      }
+    }
+
+    assignPartitions(consumer, topic, assignment);
+
+    // Seek to starting offsets.
+    for (final int partition : assignment) {
+      final long offset = nextOffsets.get(partition);
+      log.info("Seeking partition[%d] to offset[%,d].", partition, offset);
+      consumer.seek(new TopicPartition(topic, partition), offset);
+    }
+
+    return assignment;
+  }
+
+  /**
+   * Checks if the pauseRequested flag was set and if so blocks:
+   * a) if pauseMillis == PAUSE_FOREVER, until pauseRequested is cleared
+   * b) if pauseMillis != PAUSE_FOREVER, until pauseMillis elapses -or- pauseRequested is cleared
+   * <p>
+   * If pauseMillis is changed while paused, the new pause timeout will be applied. This allows adjustment of the
+   * pause timeout (making a timed pause into an indefinite pause and vice versa is valid) without having to resume
+   * and ensures that the loop continues to stay paused without ingesting any new events. You will need to signal
+   * shouldResume after adjusting pauseMillis for the new value to take effect.
+   * <p>
+   * Sets paused = true and signals paused so callers can be notified when the pause command has been accepted.
+   * <p>
+   * Additionally, pauses if all partitions assignments have been read and pauseAfterRead flag is set.
+   *
+   * @return true if a pause request was handled, false otherwise
+   */
+  private boolean possiblyPause(Set<Integer> assignment) throws InterruptedException
+  {
+    pauseLock.lockInterruptibly();
+    try {
+      if (ioConfig.isPauseAfterRead() && assignment.isEmpty()) {
+        pauseMillis = PAUSE_FOREVER;
+        pauseRequested = true;
+      }
+
+      if (pauseRequested) {
+        status = Status.PAUSED;
+        long nanos = 0;
+        hasPaused.signalAll();
+
+        while (pauseRequested) {
+          if (pauseMillis == PAUSE_FOREVER) {
+            log.info("Pausing ingestion until resumed");
+            shouldResume.await();
+          } else {
+            if (pauseMillis > 0) {
+              log.info("Pausing ingestion for %d ms", pauseMillis);
```

Comment:
`%,d` is usually easier to read for millis (it puts a comma between the seconds part and the fractional seconds part)